### PR TITLE
Introduce `--rust-edition`

### PR DESCRIPTION
**Pull request by pvdrz**
This PR introduces the concept of Rust editions to bindgen so it can generate appropriate code based on the desired edition.

The motivation behind this PR comes from #2996 which allows bindgen to generate code with C-String literals, a feature that has been available since rust 1.77 but are gated behind the 2021 edition.

The changes in this PR are:

- Introduce the `--rust-edition` CLI argument and `Builder::rust_edition` method.
- Update bindgen internals to filter features based on the set edition. 
- Gate the `literal_cstr` feature behind the 2021 edition.

cc @nyurik @emilio

### CODE REVIEW

**Comment by nyurik** [comment]
@pvdrz this is a dup of #3000 ?

**Comment by pvdrz** [comment]
> @pvdrz this is a dup of #3000 ?

yes :scream: but I cherry-picked your changes that were not included on mine and tagged you as Co-Author :sweat_smile:. Hopefully I'll coordinate this better next time

**Review by emilio — APPROVED** [review]

File: bindgen-cli/main.rs [hunk 1]
```diff
@@ -40,8 +40,13 @@ pub fn main() {
                 eprintln!("{}", info);
             }));
 
-            let bindings =
-                builder.generate().expect("Unable to generate bindings");
+            let bindings = match builder.generate() {
+                Ok(bindings) => bindings,
+                Err(err) => {
+                    eprintln!("Unable to generate bindings: {err}");
```
Comment (emilio) [inline bindgen-cli/main.rs:46]: This feels a little drive-by?

File: bindgen/features.rs [hunk 2]
```diff
@@ -38,8 +38,8 @@ impl RustTarget {
                 // fixes.
                 minor >= other_minor
             }
-            (_, Version::Nightly) => false,
             (Version::Nightly, _) => true,
+            (_, Version::Nightly) => false,
```
Comment (emilio) [inline bindgen/features.rs:42]: This too.

File: bindgen-cli/main.rs (hunk 1 above)
Comment (pvdrz) [reply bindgen-cli/main.rs:46]: I can move those to different PRs if you want

File: bindgen/features.rs (hunk 2 above)
Comment (pvdrz) [reply bindgen/features.rs:42]: Same here

**Comment by pvdrz** [comment]
Blocked by #3006
